modules/chemical_reactions/src/userobject/ThermochimicaNodalData.C

```diff
@@ -0,0 +1,259 @@
+
+#include "ThermochimicaNodalData.h"
+#include "libmesh/int_range.h"
+
+#ifdef THERMOCHIMICA_ENABLED
+#include "Thermochimica-cxx.h"
+#include "checkUnits.h"
+
+registerMooseObject("ChemicalReactionsApp", ThermochimicaNodalData);
+#endif
+
+InputParameters
+ThermochimicaNodalData::validParams()
+{
+  InputParameters params = NodalUserObject::validParams();
+  params.addClassDescription("Provides access to Thermochimica-calculated data at nodes.");
+
+  params.addRequiredCoupledVar("elements", "Amounts of elements");
+  params.addCoupledVar("element_potentials", "Chemical potentials of elements");
+
+  params.addParam<Real>("pressure", 1.0, "Pressure");
+  params.addRequiredCoupledVar("temperature", "Coupled temperature");
+
+  params.addParam<bool>("reinit_requested", true, "Should Thermochimica use re-initialization?");
+  params.addCoupledVar("output_phases", "Amounts of phases to be output");
+  params.addCoupledVar("output_species", "Amounts of species to be output");
+  return params;
+}
+
+ThermochimicaNodalData::ThermochimicaNodalData(const InputParameters & parameters)
+  : NodalUserObject(parameters),
+    _pressure(getParam<Real>("pressure")),
+    _temperature(coupledValue("temperature")),
+    _reinit_requested(getParam<bool>("reinit_requested")),
+    _n_phases(coupledComponents("output_phases")),
+    _n_species(coupledComponents("output_species")),
+    _n_elements(coupledComponents("elements")),
+    _el(_n_elements),
+    _el_name(_n_elements),
+    _ph_name(_n_phases),
+    _sp_phase_name(_n_species),
+    _sp_species_name(_n_species),
+    _output_element_potential(isCoupled("element_potentials"))
+{
+#ifndef THERMOCHIMICA_ENABLED
+  mooseError("Thermochimica disabled");
+#endif
+
+  if (n_threads() > 1)
+    mooseError("Thermochimica does not support multi-threaded runs.");
+  if (coupledComponents("element_potentials") != _n_elements)
+    paramError("element_potentials", "Specify one element potential for each entry in `elements`");
+
+  for (const auto i : make_range(_n_elements))
+  {
+    _el[i] = &coupledValue("elements", i);
+    _el_name[i] = getVar("elements", i)->name();
+  }
+
+  for (const auto i : make_range(_n_phases))
+    _ph_name[i] = getVar("output_phases", i)->name();
+
+  for (const auto i : make_range(_n_species))
+  {
+    auto species_var_name = getVar("output_species", i)->name();
+    auto semicolon = species_var_name.find(";");
+    _sp_phase_name[i] = species_var_name.substr(0, semicolon);
+    _sp_species_name[i] = species_var_name.substr(semicolon + 1);
+  }
+
+  if (_output_element_potential)
+  {
+    _element_potentials.resize(_n_elements);
+    for (const auto i : make_range(_n_elements))
+    {
+      auto element_var_name = getVar("element_potentials", i)->name();
+      auto semicolon = element_var_name.find(";");
+      _element_potentials[i] = element_var_name.substr(semicolon + 1);
+    }
+  }
+}
+
+void
+ThermochimicaNodalData::initialize()
+{
+}
+
```

Comment:
add something on mesh changed? if number of nodes changes

Reply:
or is all the initialization done in execute() ? if so add a comment

Reply:
mesh changes should not matter